File: Proyecto Mis Finanzas Java/src/baseDatos/Deserializador.java
package baseDatos;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.util.ArrayList;
import gestorAplicación.externo.Estado;
import gestorAplicación.interno.Ahorros;
import gestorAplicación.interno.Corriente;
import gestorAplicación.interno.Cuenta;
import gestorAplicación.interno.Deuda;
import gestorAplicación.interno.Metas;
import gestorAplicación.interno.Movimientos;
import gestorAplicación.interno.Usuario;
import uiMain.Main;
import gestorAplicación.externo.Banco;

public class Deserializador {
	//Deserializar Arraylists con objetos
	@SuppressWarnings("unchecked")
	public static String deserializar_listas(String clase) {		
		switch(clase) {
			case "Usuarios":
				try{
					File f = new File("");
					FileInputStream fe = new FileInputStream(new File(f.getAbsolutePath() + "\\src\\baseDatos\\temp\\" + Usuario.nombreD + "_lista" + ".dat"));
					ObjectInputStream streamEntrada = new ObjectInputStream(fe);
					Deserializador.readObject(streamEntrada, clase);
					streamEntrada.close();
					return ("Lista con " + Usuario.getUsuariosTotales().size() + " usuarios ha sido cargada con éxito");
			
				}catch(IOException ex) {
					return ("La lista de Usuarios no pudo ser deserializada en el sistema: " + ex);
				} catch (ClassNotFoundException ex) {
					return ("La lista de Usuarios no pudo ser deserializada en el sistema: " + ex);
				}				
			case "Estados":
				try{
					File f = new File("");
					FileInputStream fe = new FileInputStream(new File(f.getAbsolutePath() + "\\src\\baseDatos\\temp\\" + Estado.nombreD + "_lista" + ".dat"));
					ObjectInputStream streamEntrada = new ObjectInputStream(fe);
					Deserializador.readObject(streamEntrada, clase);
					streamEntrada.close();
					return ("Lista con " + Estado.getEstadosTotales().size() + " estados ha sido cargada con éxito");
				
				}catch(IOException ex) {
					return ("La lista de Estados no pudo ser deserializada en el sistema: " + ex);
				} catch (ClassNotFoundException ex) {
					return ("La lista de Estados no pudo ser deserializada en el sistema: " + ex);
				}	
			case "Bancos":
				try{
					File f = new File("");
					FileInputStream fe = new FileInputStream(new File(f.getAbsolutePath() + "\\src\\baseDatos\\temp\\" + Banco.nombreD + "_lista" + ".dat"));
					ObjectInputStream streamEntrada = new ObjectInputStream(fe);
					Deserializador.readObject(streamEntrada, clase);
					streamEntrada.close();
					return ("Lista con " + Banco.getBancosTotales().size() + " bancos ha sido cargada con éxito");				
				}catch(IOException ex) {
					return ("La lista de Estados no pudo ser deserializada en el sistema: " + ex);
				} catch (ClassNotFoundException ex) {
					return ("La lista de Estados no pudo ser deserializada en el sistema: " + ex);
				}	
			case "Cuentas":
				try{
					File f = new File("");
					FileInputStream fe = new FileInputStream(new File(f.getAbsolutePath() + "\\src\\baseDatos\\temp\\" + Cuenta.nombreD + "_lista" + ".dat"));
					ObjectInputStream streamEntrada = new ObjectInputStream(fe);
					Deserializador.readObject(streamEntrada, clase);
					streamEntrada.close();
					return ("Lista con " + Cuenta.getCuentasTotales().size() + " cuentas ha sido cargada con éxito");				
				}catch(IOException ex) {
					return ("La lista de Cuentas no pudo ser deserializada en el sistema: " + ex);
				} catch (ClassNotFoundException ex) {
					return ("La lista de Cuentas no pudo ser deserializada en el sistema: " + ex);
				}	
			case "Movimientos":
				try{
					File f = new File("");
					FileInputStream fe = new FileInputStream(new File(f.getAbsolutePath() + "\\src\\baseDatos\\temp\\" + Movimientos.nombreD + "_lista" + ".dat"));
					ObjectInputStream streamEntrada = new ObjectInputStream(fe);
					Deserializador.readObject(streamEntrada, clase);
					streamEntrada.close();
					return ("Lista con " + Movimientos.getMovimientosTotales().size() + " movimientos ha sido cargada con éxito");				
				}catch(IOException ex) {
					return ("La lista de Movimientos no pudo ser deserializada en el sistema: " + ex);
				} catch (ClassNotFoundException ex) {
					return ("La lista de Movimientos no pudo ser deserializada en el sistema: " + ex);
				}			
			case "Metas":
				try{
					File f = new File("");
					FileInputStream fe = new FileInputStream(new File(f.getAbsolutePath() + "\\src\\baseDatos\\temp\\" + Metas.nombreD + "_lista" + ".dat"));
					ObjectInputStream streamEntrada = new ObjectInputStream(fe);
					Deserializador.readObject(streamEntrada, clase);
					streamEntrada.close();
					return ("Lista con " + Metas.getMetasTotales().size() + " metas ha sido cargada con éxito");				
				}catch(IOException ex) {
					return ("La lista de Metas no pudo ser deserializada en el sistema: " + ex);
				} catch (ClassNotFoundException ex) {
					return ("La lista de Metas no pudo ser deserializada en el sistema: " + ex);
				}		
			default:
				return("Error de deserialización: El objeto debe estar definido en el sistema.");
		}
	}
	
	private static void readObject(ObjectInputStream in, String clase) throws IOException, ClassNotFoundException {
	    switch(clase) {
		    case "Usuarios":
		    	Usuario.setUsuariosTotales((ArrayList<Usuario>) in.readObject());
		    	for(Usuario u : Usuario.getUsuariosTotales()) {
		    		if(u.getNombre().equals("admin")) {
		    			Main.setContraseñaAdmin("admin");
		    		}	
		    	}
		    	break;
		    case "Cuentas":
		    	Cuenta.setCuentasTotales((ArrayList<Cuenta>) in.readObject());
		    	for(Cuenta c : Cuenta.getCuentasTotales()) {
		    		if(c instanceof Ahorros) {
		    			Ahorros.getCuentasAhorroTotales().add((Ahorros) c);
		    		}else {
		    			Corriente.getCuentasCorrienteTotales().add((Corriente) c);
		    		}
		    	}
		    	break;
		    case "Bancos":
		    	Banco.setBancosTotales((ArrayList<Banco>) in.readObject());
		    	break;
		    case "Estados":
		    	Estado.setEstadosTotales((ArrayList<Estado>) in.readObject());
		    	break;
		    case "Metas":
		    	Metas.setMetasTotales((ArrayList<Metas>) in.readObject());
		    	for(Metas m : Metas.getMetasTotales()) {
		    		if(m instanceof Deuda) {
		    			Deuda.getDeudasTotales().add((Deuda) m);
		    		}
		    	}
		    	break;
		    case "Movimientos":
		    	Movimientos.setMovimientosTotales((ArrayList<Movimientos>) in.readObject());
		    	break;
		}		
	}
}

File: Proyecto Mis Finanzas Python/src/baseDatos/deserializador.py
import pickle as pk
import os
import os.path as os_query
from gestorAplicación.interno.usuario import Usuario
from gestorAplicación.interno.movimientos import Movimientos
from gestorAplicación.externo.estado import Estado
from gestorAplicación.externo.banco import Banco
from gestorAplicación.interno.ahorros import Ahorros, Cuenta
from gestorAplicación.interno.corriente import Corriente
from gestorAplicación.interno.deuda import Deuda, Metas

class Deserializador():
    # Método para deserializar las listas con objetos de cualquier clase, que añade automáticamente los objetos a las listas de clase respectiva para cada clase
    
    # Obtener ruta relativa
    current_directory = os.path.dirname(os.path.abspath(__file__))
    route_db = os.path.join(current_directory + '\\temp')

    @classmethod
    def deserializar(cls, clase) -> list:
        if(clase == "Usuarios"):
            if(os_query.isfile(cls.route_db + "\lista_usuarios.pkl")):
                objects_file = open(cls.route_db + "\lista_usuarios.pkl", "rb")
                objects_list = pk.load(objects_file)
                objects_file.close()
                Usuario.setUsuariosTotales(objects_list)
        if(clase == "Estados"):
            if(os_query.isfile(cls.route_db + "/lista_estados.pkl")):
                objects_file = open(cls.route_db + "/lista_estados.pkl", "rb")
                objects_list = pk.load(objects_file)
                objects_file.close()
                Estado.setEstadosTotales(objects_list)
        if(clase == "Bancos"):
            if(os_query.isfile(cls.route_db + "/lista_bancos.pkl")):
                objects_file = open(cls.route_db + "/lista_bancos.pkl", "rb")
                objects_list = pk.load(objects_file)
                objects_file.close()
                Banco.setBancosTotales(objects_list)
        if(clase == "Cuentas"):
            if(os_query.isfile(cls.route_db + "/lista_cuentas.pkl")):
                objects_file = open(cls.route_db + "/lista_cuentas.pkl", "rb")
                objects_list = pk.load(objects_file)
                objects_file.close()
                Cuenta.setCuentasTotales(objects_list)
                for account in Cuenta.getCuentasTotales():
                    if(isinstance(account, Ahorros)):
                        Ahorros.getCuentasAhorrosTotales().append(account)
                    else:
                        Corriente.getCuentasCorrienteTotales().append(account)
        if(clase == "Movimientos"):
            if(os_query.isfile(cls.route_db + "/lista_movimientos.pkl")):
                objects_file = open(cls.route_db + "/lista_movimientos.pkl", "rb")
                objects_list = pk.load(objects_file)
                objects_file.close()
                Movimientos.setMovimientosTotales(objects_list)
        if(clase == "Metas"):
            if(os_query.isfile(cls.route_db + "/lista_metas.pkl")):
                objects_file = open(cls.route_db + "/lista_metas.pkl", "rb")
                objects_list = pk.load(objects_file)
                objects_file.close()
                Metas.setMetasTotales(objects_list)
                for meta in Metas.getMetasTotales():
                    if(isinstance(meta, Deuda)):
                        Deuda.getDeudasTotales().append(meta)
